### PR TITLE
Convert check_security to a PHP script

### PR DESCRIPTION
**Pull request by cpeel**
Convert `check_security.sh` to a PHP script and have it check for system calls as well. Include an approved-list for the current uses we hope to change later.

This is infra in support of https://github.com/DistributedProofreaders/dproofreaders/issues/1274

### CODE REVIEW

File: SETUP/ci/check_security.php [hunk 1]
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env php
+<?php
+// Check all .php and .inc files for security best-practices.
+
+$relPath = "../../pinc/";
+include_once($relPath."misc.inc");
```
[the remark this answers is not in the document]
Comment (cpeel) [reply SETUP/ci/check_security.php:6]: That's true for all site pages -- where we need a database connection and gettext set up, etc -- but not for most things in `SETUP/`.

File: SETUP/ci/check_security.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    // No file should include a system call (use Symfony Process instead)
+    if (file_includes_system_call("$basedir/$file") && !in_array($file, $ok_system_calls)) {
+        abort("file includes system(), exec(), passthru(), shell_exec(), or escapeshellcmd()");
+    }
+
+}
+
+function get_all_php_files($basedir)
```
[the remark this answers is not in the document]
Comment (cpeel) [reply SETUP/ci/check_security.php:72]: Good all and thanks, I copy/pastad this from an existing script (bad me). I'll get these fixed.

File: SETUP/ci/check_security.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+function file_includes_system_call($filename)
+{
+    // It's easy to get system calls wrong so prevent adding more, preferring
+    // to use Symfony Process() instead. We don't really care about
+    // escapeshellcmd() but it's a sign that there's probably some system/shell
+    // magic happening that we want to catch.
+
+    $contents = file_get_contents($filename);
+    return preg_match("/\b(?:exec|system|passthru|shell_exec|escapeshellcmd)\(/", $contents, $matches);
```
[the remark this answers is not in the document]
Comment (cpeel) [reply SETUP/ci/check_security.php:106]: We only check PHP code so I'm not worried about HTML/JS/CSS. A real parser would be better but seems overkill for this particular test.

File: SETUP/ci/check_security.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+// List of files that can contain mysqli_error() calls
+$ok_mysqli_error_calls = [
+    "pinc/DPDatabase.inc",
+];
+
+$basedir .= endswith($basedir, "/") ? "" : "/";
+$files = get_all_php_files($basedir);
+foreach ($files as $file) {
+    // If it's in the SETUP directory, skip it
+    if (startswith($file, "SETUP/")) {
```
Comment (chrismiceli) [inline SETUP/ci/check_security.php:39]: would it be simpler to have an array of files startswith to skip?

**Review by chrismiceli — APPROVED** [review]

File: SETUP/ci/check_security.php (hunk 4 above)
Comment (cpeel) [reply SETUP/ci/check_security.php]: Yeah, and it's not as though the comments are adding anything to those if statements either. Done.

**Review by chrismiceli — APPROVED** [review]

**Review by srjfoo — APPROVED** [review]

File: SETUP/ci/check_security.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (jchaffraix) [reply SETUP/ci/check_security.php:101]: Makes sense, I expected something like this. Thanks for confirming!

**Review by jchaffraix — APPROVED** [review]

**Pull request merged**